src/Microsoft.AspNetCore.Routing.Abstractions/RouteData.cs

```diff
@@ -222,15 +234,41 @@ public void Restore()
                 }
                 else if (_routers == null)
                 {
-                    _routeData._routers.Clear();
+                    // Perf: this is optimized for small list sizes, in particular to avoid overhead of a native call in
+                    // Array.Clear inside the List.Clear() method.
+                    var routers = _routeData._routers;
+                    for (var i = routers.Count - 1; i >=0 ; i--)
+                    {
+                        routers.RemoveAt(i);
+                    }
                 }
                 else
                 {
-                    _routeData._routers.Clear();
+                    // Perf: this is optimized for small list sizes, in particular to avoid overhead of a native call in
+                    // Array.Clear inside the List.Clear() method.
+                    //
+                    // We want to basically copy the contents of _routers in _routeData._routers - this change does
+                    // that with the minimal number of reads/writes and without calling Clear().
+                    var routers = _routeData._routers;
+                    var snapshotRouters = _routers;
+
+                    // This is made more complicated by the fact that List[int] throws if i == Count, so we have
+                    // to do two loops and call Add for those cases.
+                    var i = 0;
+                    for (; i < snapshotRouters.Count && i < routers.Count; i++)
+                    {
+                        routers[i] = snapshotRouters[i];
+                    }
+
+                    for (; i < snapshotRouters.Count; i++)
+                    {
+                        routers.Add(snapshotRouters[i]);
+                    }
 
-                    for (var i = 0; i < _routers.Count; i++)
+                    // Trim excess
+                    for (i = routers.Count - 1; i >= snapshotRouters.Count; i--)
```

Comment:
Can you add a comment here stating `RemoveRange` has the same overhead as `Clear`?